ls1intum/Artemis
# [Feature] Grading System - Absolute Points for Grade Key

### Checklist

### Motivation and Context


Previously, instructors could only create a grade key by setting the boundaries of the grade steps in percentages. For convenience now, they can also set the grade key using absolute points, as long as the max points property for the course/exam has been set.

### Description

I've created a max points field for the course which servers a similar purpose as the one for the exam and is necessary for the other changes in this PR.
I've added a new input field on the grade key page, which sets and displays the max points for courses/exams. Furthermore, the grade step table has 2 additional columns for the minimum and maximum points. Changing either of the values of a boundary (the percentage or the points) automatically adapts the other value. Also, changing the max points of the course/exam maintains the percentage values, but adapts the absolute points values accordingly. 
It's worth mentioning that the grade key matching still relies solely on the percentage values, and the possibility to add the steps with absolute points is only there for convenience.

### Steps for Testing


1. Log in to Artemis as an Instructor (Using TS3 and my course "IvoTC" could save you some time, I have an exam there with the necessary student submissions for testing)
2. Navigate to Course Administration -> Grade Key
3. Try entering different combinations of values.
- if max points for the course aren't set, the min/max points fields should be empty and disabled, but valid grade keys should still be saveable
- updating the points/percentages for a lower/upper bound should update the percentages/points for that same bound, such that (percentage * max points) / 100 = points always holds (with an epsilon margin of error)
- updating the max points, should automatically adapt all grade steps, where the percentages are kept, and the points are adjusted accordingly
4. Go to the course edit page -> there should be a warning next to the max points fields regarding changing the value
5. Try updating the value and saving it -> then go back to the grade key page -> the value should be adjusted to the one in the edit page and all of the grade steps should have been adjusted as in step 3 (percentages maintained/points adapted)
6. Repeat the steps accordingly for an exam and its grade key
7. Finally, find an exam with student submissions and published results and create a grade key for it. Then log in as one of the students with submission in this exam and go to the exam overview page -> Click on View Grade Key -> the intervals should be displayed in points as well.


### Test Coverage



GradingScaleService.java: 94%
GradingScaleRepository.java: 100%
grade-key-overview.component.ts: 97.22%
grading-system.component.ts: 94.25%


### Screenshots


Grade key page:
![Screenshot 2021-06-14 162937](https://user-images.githubusercontent.com/63168603/121900122-bd6a1300-cd2d-11eb-99dc-31f8be032a93.png)
Course/Exam edit page:
![Screenshot 2021-06-14 163018](https://user-images.githubusercontent.com/63168603/121900189-d4a90080-cd2d-11eb-9ae0-70975fe94aa2.png)
Grade Key Overview for exams:
![image](https://user-images.githubusercontent.com/63168603/121900291-f30efc00-cd2d-11eb-9eb3-175a36b28406.png)
